flight/src/conversion.rs

```diff
@@ -0,0 +1,159 @@
+use crate::types::*;
+use anyhow::Result;
+use arrow::array::{
+    ArrayRef, BinaryBuilder, ListBuilder, PrimitiveBuilder, StringBuilder, StructArray,
+    UInt8Builder,
+};
+use arrow::datatypes::{DataType, Field};
+use arrow::record_batch::RecordBatch;
+use std::sync::Arc;
+
+#[derive(Debug)]
+pub struct ConclusionEventBuilder {
+    event_type: UInt8Builder,
+    stream_id: BinaryBuilder,
+    stream_type: UInt8Builder,
+    controller: StringBuilder,
+    data: BinaryBuilder,
+    previous: ListBuilder<BinaryBuilder>,
+}
+
+impl Default for ConclusionEventBuilder {
+    fn default() -> Self {
+        Self {
+            event_type: PrimitiveBuilder::new(),
+            stream_id: BinaryBuilder::new(),
+            stream_type: PrimitiveBuilder::new(),
+            controller: StringBuilder::new(),
+            data: BinaryBuilder::new(),
+            previous: ListBuilder::new(BinaryBuilder::new())
+                .with_field(Field::new_list_field(DataType::Binary, false)),
+        }
+    }
+}
+
+impl ConclusionEventBuilder {
+    fn append(&mut self, event: &ConclusionEvent) {
+        match event {
+            ConclusionEvent::Data(data_event) => {
+                self.event_type.append_value(0);
+                self.stream_id.append_value(data_event.id.to_bytes());
+                self.stream_type.append_value(data_event.init.stream_type);
+                self.controller.append_value(&data_event.init.controller);
+                self.data.append_value(&data_event.data);
+                for cid in &data_event.previous {
+                    self.previous.values().append_value(cid.to_bytes());
+                }
+                self.previous.append(true);
+            }
+            ConclusionEvent::Time(_) => {
+                todo!("implement time event once we know what its structure looks like");
+            }
+        }
+    }
+
+    fn finish(&mut self) -> StructArray {
+        let event_type = Arc::new(self.event_type.finish()) as ArrayRef;
+        let event_type_field = Arc::new(Field::new("event_type", DataType::UInt8, false));
+
+        let stream_id = Arc::new(self.stream_id.finish()) as ArrayRef;
+        let stream_id_field = Arc::new(Field::new("stream_id", DataType::Binary, false));
+
+        let stream_type = Arc::new(self.stream_type.finish()) as ArrayRef;
+        let stream_type_field = Arc::new(Field::new("stream_type", DataType::UInt8, false));
+
+        let controller = Arc::new(self.controller.finish()) as ArrayRef;
+        let controller_field = Arc::new(Field::new("controller", DataType::Utf8, false));
+
+        let data = Arc::new(self.data.finish()) as ArrayRef;
+        let data_field = Arc::new(Field::new("data", DataType::Binary, false));
```

Comment:
The `data` field will need to be nullable because time events will not have data values.